### DISCUSSION TITLE
Missing styles after Pretix update from source (development version installation) from v2024.1.0 to v2024.9.0

### DISCUSSION THREAD

**Discussion by c0de-bender** [pretix/pretix#4548] [Q&A]
Hi,

I installed a pretix development version from source last year and now tried to update it to a newer version. The old version was based on Pretix `v2024.1.0`, the new one after the update on `v2024.9.0`.

In principal, the update worked, but the styles are missing afterwards. My installation from source involves building Pretix first, then installing the build and executing the usual Pretix update steps. The update seems to go fine – I don't see errors or warnings in the logs that I haven't seen in previous updates. But when I open Pretix in the browser, I just see plain html without any styles.

In the Pretix logs after startup I can see some error messages which state that in fact static files are missing (see below for the detailed messages).

Is there something that has changed since version `v2024.1.0` in the way how  to build Pretix? E.g. I saw that you made some changes to your `.gitlab-ci.yml` in [this commit](https://github.com/pretix/pretix/commit/8010d2e6bbefbc15c7abecddcbf69d41011d76ef). Is there something there that has changed that I would also need to adjust?

I tried to research more potential causes, but found nothing that would hint me what the problem is. I'm a bit lost, especially because I don't see any error in the update logs.

**Does anyone have any idea what the problem might be and how I could fix it? I appreciate any hint about the cause, or what to investigate / check, and about potential solutions!**

**Note 1**: I saw that the command `python3.11 -m pretix updatestyles` changed to `python3.11 -m pretix updateassets` and adjusted that in my script. I also tried with the old `updatestyles` and that gave the same result (missing styles).

**Note 2**: I also saw the hint on the [Dealing with errors page](https://docs.pretix.eu/en/latest/admin/errors.html) about missing styles (update nginx config after OS's Python version update), but I guess (or hope) that this is not the problem, because I'm still using `python3.11` as before. It's a different minor version though (3.11.10 instead of previously 3.11.5). And I also don't have access to the webserver config as I'm on shared hosting.
If you think that this is still one likely issue, I could ask my hoster about it.

**Note 3**: I also see some warning messages in the build step log output, but I checked my old update logs and had the same warnings in February 2024 when I last successfully updated Pretix. Therefore I guess that these are not the issue. I appended them anyway, just in case. See below "Warnings during build step".

Below I share more information about my setup, update procedure, version information and error and warning messages.

Thanks for reading!

## Pretix version information

I made only very small changes to the Pretix sources (a few translation changes in one translation file and replacing `class="panel panel-default ..."` by `class="panel panel-primary ..."` in two html files). Apart from that I'm trying to update from source with these versions:

Pretix version before update: `v2024.1.0`
Pretix version after update: `v2024.9.0`

## System information

OS: CentOS `7` (shared hosting)

Python `3.11.10`
pip `24.2`
npm `10.7.0`
node `18.20.4`

After the upgrading commands in the update procedure:

setuptools `75.2.0`
wheel `0.44.0`
gunicorn `23.0.0`
build `1.2.2.post1`

## My update commands

I had made a custom installation of Pretix from source last year. I created it by following the [Pretix Manual installation guide](https://docs.pretix.eu/en/latest/admin/installation/manual_smallscale.html) and the [Pretix guide for installing a development version](https://docs.pretix.eu/en/latest/admin/installation/dev_version.html#manual-installation) and adapting them.

I also adapted the update procedure accordingly. In both cases, I'm building Pretix myself as a bdist from source (the direct installation from source as described in the [Pretix guide for installing a development version](https://docs.pretix.eu/en/latest/admin/installation/dev_version.html#manual-installation) didn't work on my system, see [my problems with this last year](https://github.com/pretix/pretix/discussions/3642)).

This is my update script:

```sh

PRETIX_SOURCE_TREE_ROOT=~/pretix/pretix-src-repo

# Stop the Pretix services
supervisorctl stop pretix pretix_worker

## Update Pretix
# Update/Install Python tools and gunicorn webserver
pip3.11 install --user -U --upgrade-strategy only-if-needed pip setuptools wheel gunicorn

# Update/Install Python package "build"
pip3.11 install --user -U --upgrade-strategy only-if-needed build

cd $PRETIX_SOURCE_TREE_ROOT

# Clean the dist directory, just in case it's not empty ...
rm -rf dist/*

# Build Pretix (as a binary wheel-dist)
python3.11 -m build --wheel

# Install the binary wheel-dist
echo "Installing the bdist via pip to ~/.local/lib"
#Use --force-reinstall, if Pretix was installed beforehand with the same version! Otherwise, the installation step is skipped!
pip3.11 install --user --force-reinstall -U ${PRETIX_SOURCE_TREE_ROOT}/dist/pretix-*.whl

# Migrate the database
python3.11 -m pretix migrate

# Rebuild Pretix static files
python3.11 -m pretix rebuild

# Update assets
python3.11 -m pretix updateassets

# Restart the Pretix services
supervisorctl start pretix pretix_worker

```

## Error messages in Pretix logs after startup

In my supervisord logs I see error messages like these:

```sh
2024-10-22 21:52:38,370 INFO success: pretix_worker entered RUNNING state, process has stayed up for > than 30 seconds (startsecs)
2024-10-22 21:52:45,601 DEBG 'pretix' stderr output:
WARNING 2024-10-22 21:52:45,598 django.request log Not Found: /static/CACHE/js/output.97119b17bde0.js
WARNING 2024-10-22 21:52:45,598 django.request log Not Found: /static/CACHE/js/output.97119b17bde0.js

2024-10-22 21:52:47,550 DEBG 'pretix' stderr output:
WARNING 2024-10-22 21:52:47,549 django.request log Not Found: /static/CACHE/css/output.20975e6a0b0e.css
WARNING 2024-10-22 21:52:47,549 django.request log Not Found: /static/CACHE/css/output.20975e6a0b0e.css

2024-10-22 21:52:47,706 DEBG 'pretix' stderr output:
WARNING 2024-10-22 21:52:47,705 django.request log Not Found: /static/CACHE/js/output.97119b17bde0.js
WARNING 2024-10-22 21:52:47,705 django.request log Not Found: /static/CACHE/js/output.97119b17bde0.js

2024-10-22 21:52:53,563 DEBG 'pretix' stderr output:
WARNING 2024-10-22 21:52:53,563 django.request log Not Found: /static/CACHE/css/output.e9e25fb92801.css
WARNING 2024-10-22 21:52:53,563 django.request log Not Found: /static/CACHE/css/output.e9e25fb92801.css

2024-10-22 21:52:53,564 DEBG 'pretix' stderr output:
WARNING 2024-10-22 21:52:53,563 django.request log Not Found: /static/pretixcontrol/js/auth.041e785254b9.js
WARNING 2024-10-22 21:52:53,563 django.request log Not Found: /static/pretixcontrol/js/auth.041e785254b9.js

2024-10-22 21:52:57,916 DEBG 'pretix' stderr output:
WARNING 2024-10-22 21:52:57,916 django.request log Not Found: /static/CACHE/css/output.20975e6a0b0e.css
WARNING 2024-10-22 21:52:57,916 django.request log Not Found: /static/CACHE/css/output.20975e6a0b0e.css

2024-10-22 21:52:57,940 DEBG 'pretix' stderr output:
WARNING 2024-10-22 21:52:57,939 django.request log Not Found: /static/CACHE/js/output.97119b17bde0.js
WARNING 2024-10-22 21:52:57,939 django.request log Not Found: /static/CACHE/js/output.97119b17bde0.js

2024-10-22 21:53:12,577 DEBG 'pretix' stderr output:
WARNING 2024-10-22 21:53:12,576 django.request log Not Found: /static/CACHE/css/output.e9e25fb92801.css
WARNING 2024-10-22 21:53:12,576 django.request log Not Found: /static/CACHE/css/output.e9e25fb92801.css

2024-10-22 21:53:12,578 DEBG 'pretix' stderr output:
WARNING 2024-10-22 21:53:12,577 django.request log Not Found: /static/pretixcontrol/js/auth.041e785254b9.js
WARNING 2024-10-22 21:53:12,577 django.request log Not Found: /static/pretixcontrol/js/auth.041e785254b9.js

2024-10-22 21:53:14,818 DEBG 'pretix' stderr output:
WARNING 2024-10-22 21:53:14,817 django.request log Not Found: /static/pretixbase/img/pretix-logo.c1443d5b8a38.svg
WARNING 2024-10-22 21:53:14,817 django.request log Not Found: /static/pretixbase/img/pretix-logo.c1443d5b8a38.svg

[...]
```

There are more lines like this.

## Warnings during build step

I mentioned some warnings in the build step that I get, but I think are unrelated, because I had them before with successful updates too. Here are they:

My build command (see above) is:

`python3.11 -m build --wheel`

and the warnings that I get look like this:

```sh
/tmp/build-env-kpt00dql/lib64/python3.11/site-packages/setuptools/command/build_py.py:218: _Warning: Package 'pretix.base.templates' is absent from the `packages` configuration.
```

Here is the longer output with the most important warning parts (as far as I can see) of the log output of this command:

```sh
* Creating isolated environment: venv+pip...
* Installing packages in isolated environment:
  - importlib_metadata
  - setuptools
  - setuptools-rust
  - tomli
  - wheel
* Getting build dependencies for wheel...
running egg_info
writing src/pretix.egg-info/PKG-INFO
writing dependency_links to src/pretix.egg-info/dependency_links.txt
writing entry points to src/pretix.egg-info/entry_points.txt
writing requirements to src/pretix.egg-info/requires.txt
writing top-level names to src/pretix.egg-info/top_level.txt
reading manifest file 'src/pretix.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching '*' under directory 'src/pretix/static.dist'
warning: no files found matching '*' under directory 'src/pretix/helpers/locale'
warning: no files found matching '*' under directory 'src/pretix/plugins/autocheckin/static'
warning: no files found matching '*' under directory 'src/pretix/plugins/manualpayment/templates'
warning: no files found matching '*' under directory 'src/pretix/plugins/manualpayment/static'
warning: no files found matching '*' under directory 'src/pretix/plugins/src/pretixdroid/templates'
warning: no files found matching '*' under directory 'src/pretix/plugins/src/pretixdroid/static'
warning: no files found matching '*' under directory 'src/pretix/plugins/ticketoutputpdf/static'
warning: no files found matching '*' under directory 'src/pretix/plugins/returnurl/static'
warning: no previously-included files matching '*' found under directory 'doc'
warning: no previously-included files matching '*' found under directory 'deployment'
warning: no previously-included files matching '*' found under directory 'res'
adding license file 'LICENSE'
writing manifest file 'src/pretix.egg-info/SOURCES.txt'
* Installing packages in isolated environment:
  - BeautifulSoup4==4.12.*
  - Django[argon2]==4.2.*,>=4.2.15
  - Pillow==10.4.*

[...]

running bdist_wheel
running build
npm warn deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
npm warn deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm warn deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm warn deprecated source-map-resolve@0.5.3: See https://github.com/lydell/source-map-resolve#deprecated
npm warn deprecated querystring@0.2.1: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
npm warn deprecated vue@2.7.16: Vue 2 has reached EOL and is no longer actively maintained. See https://v2.vuejs.org/eol/ for more details.

added 346 packages, and audited 347 packages in 9s

[...]

processing language tr
processing language uk

963 static files copied to '/home/<user>/pretix/pretix-src-repo/src/pretix/static.dist', 963 post-processed.
Compressing... done
Compressed 30 block(s) from 607 template(s) for 1 context(s).
running build_py
creating build/lib/pretix
copying src/pretix/__init__.py -> build/lib/pretix

[...]

creating build/lib/pretix/presale/management/commands
copying src/pretix/presale/management/commands/__init__.py -> build/lib/pretix/presale/management/commands
copying src/pretix/presale/management/commands/updatestyles.py -> build/lib/pretix/presale/management/commands
copying src/pretix/presale/management/commands/updateassets.py -> build/lib/pretix/presale/management/commands
running egg_info
writing src/pretix.egg-info/PKG-INFO
writing dependency_links to src/pretix.egg-info/dependency_links.txt
writing entry points to src/pretix.egg-info/entry_points.txt
writing requirements to src/pretix.egg-info/requires.txt
writing top-level names to src/pretix.egg-info/top_level.txt
reading manifest file 'src/pretix.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching '*' under directory 'src/pretix/helpers/locale'
warning: no files found matching '*' under directory 'src/pretix/plugins/autocheckin/static'
warning: no files found matching '*' under directory 'src/pretix/plugins/manualpayment/templates'
warning: no files found matching '*' under directory 'src/pretix/plugins/manualpayment/static'
warning: no files found matching '*' under directory 'src/pretix/plugins/src/pretixdroid/templates'
warning: no files found matching '*' under directory 'src/pretix/plugins/src/pretixdroid/static'
warning: no files found matching '*' under directory 'src/pretix/plugins/ticketoutputpdf/static'
warning: no files found matching '*' under directory 'src/pretix/plugins/returnurl/static'
warning: no previously-included files matching '*' found under directory 'doc'
warning: no previously-included files matching '*' found under directory 'deployment'
warning: no previously-included files matching '*' found under directory 'res'
adding license file 'LICENSE'
writing manifest file 'src/pretix.egg-info/SOURCES.txt'
/tmp/build-env-kpt00dql/lib64/python3.11/site-packages/setuptools/command/build_py.py:218: _Warning: Package 'pretix.base.templates' is absent from the `packages` configuration.
!!

        ********************************************************************************
        ############################
        # Package would be ignored #
        ############################
        Python recognizes 'pretix.base.templates' as an importable package[^1],
        but it is absent from setuptools' `packages` configuration.

        This leads to an ambiguous overall configuration. If you want to distribute this
        package, please make sure that 'pretix.base.templates' is explicitly added
        to the `packages` configuration field.

        Alternatively, you can also rely on setuptools' discovery methods
        (for example by using `find_namespace_packages(...)`/`find_namespace:`
        instead of `find_packages(...)`/`find:`).

        You can read more about "package discovery" on setuptools documentation page:

        - https://setuptools.pypa.io/en/latest/userguide/package_discovery.html

        If you don't want 'pretix.base.templates' to be distributed and are
        already explicitly excluding 'pretix.base.templates' via
        `find_namespace_packages(...)/find_namespace` or `find_packages(...)/find`,
        you can try to use `exclude_package_data`, or `include-package-data=False` in
        combination with a more fine grained `package-data` configuration.

        You can read more about "package data files" on setuptools documentation page:

        - https://setuptools.pypa.io/en/latest/userguide/datafiles.html


        [^1]: For Python, any directory (with suitable naming) can be imported,
              even if it does not contain any `.py` files.
              On the other hand, currently there is no concept of package data
              directory, all directories are treated like packages.
        ********************************************************************************

!!
  check.warn(importable)
/tmp/build-env-kpt00dql/lib64/python3.11/site-packages/setuptools/command/build_py.py:218: _Warning: Package 'pretix.base.templates.bootstrap3' is absent from the `packages` configuration.
!!

[...]

```

And then it repeates a lot of pretix packages with the same warning message.

As I said before, I had these warnings during the last successful update in February too. So my guess is that these warnings are not related to the missing styles. But they seem like they shouldn't be there anyway (and I don't see any other problems in the log output of my update commands).
Do you think they might still be related to missing styles?
Even if they are unrelated, do you have an idea why they occur and how to get rid of them?

**Comment by felixrindt** [comment]
Are you sure this is not the same issue as #4432 ?

**Reply by c0de-bender** [reply]
Hey, thanks a lot for your reply!

It's very possible that this is the same issue. I didn't find https://github.com/pretix/pretix/issues/4432 before when I was searching for similar problems.

I don't understand yet how to fix it from the thread there though.

I'm also on uberspace (as you are). Did you find a way to fix it? If so, how did you fix the issue?

If not, I'd just roll back to 2024.1.0 with my backup and then only update to 2024.7.x to avoid that issue.

**Reply by felixrindt** [reply]
I'm still on 2024.07. because I did not get around to fixing my setup yet.

**Reply by felixrindt** [reply]
In case you wanna try @c0de-bender:
pretix (like any django project) collects its static files in a "staticfiles" folder. It tells you in the build step, should be something like `/home/jupretix/.local/lib/python3.11/site-packages/pretix/static.dist'` :)
Using uberspace [web backends](https://manual.uberspace.de/web-backends/#specific-path) you should be able to point their apache to serve those static files directly (that's what pretix has been doing for you before 2024.10 even though it was not officially supported). Feel free to report what worked for you.

**Reply by c0de-bender** [reply]
@felixrindt Thanks for the hint! Maybe I'll try that at a later point (don't have a lot of time at the moment).

For now I just downgraded to v2024.1.0 after I restored my backup, and then updated to v2024.7.1. That has worked and I'm satisfied with that for now for the project that I'm working on.

**Reply by felixrindt** [reply]
I opened a [PR over at uberspace lab](https://github.com/Uberspace/lab/pull/1828) to include serving static files. Should be straight forward to fix your setup. Feel free to try.
